File: 02. DurableConsumptionModel/cppfuncs/trans.cpp
#ifndef MAIN
#define TRANS
#include "header.cpp"
#endif

namespace trans {
    
double p_plus_func(double p, double psi, par_struct* par){
    double p_plus = p*psi;
    p_plus = MAX(p_plus,par->p_min); 
    p_plus = MIN(p_plus,par->p_max); 
    return p_plus;
}

double n_plus_func(double d, par_struct* par){
    double n_plus = (1.0-par->delta)*d;
    n_plus = MIN(n_plus,par->n_max);
    return n_plus;
}

double n1_plus_func(double d1, par_struct* par){
    double n1_plus = (1.0-par->delta1)*d1;
    n1_plus = MIN(n1_plus,par->n_max);
    return n1_plus;
}

double n2_plus_func(double d2, par_struct* par){
    double n2_plus = (1.0-par->delta2)*d2;
    n2_plus = MIN(n2_plus,par->n_max);
    return n2_plus;
}

double m_plus_func(double a, double p_plus, double xi_plus, par_struct* par){
    double y_plus = p_plus*xi_plus;
    double m_plus = par->R*a+ y_plus;
    return m_plus;

}

double x_plus_func(double m_plus, double n_plus, par_struct* par){
    return m_plus + (1.0-par->tau)*n_plus;
}

double x_plus_full_func(double m_plus, double n1_plus, double n2_plus, par_struct* par){
    return m_plus + (1.0-par->tau1)*n1_plus + (1.0-par->tau2)*n2_plus;
}

double x_plus_d1_func(double m_plus, double n1_plus, par_struct* par){
    return m_plus + (1.0-par->tau1)*n1_plus;
}

double x_plus_d2_func(double m_plus, double n2_plus, par_struct* par){
    return m_plus + (1.0-par->tau2)*n2_plus;
}


} // namespace

File: 02. DurableConsumptionModel/trans.py
import numpy as np
from numba import njit

@njit(fastmath=True)
def p_plus_func(p,psi,par):
    p_plus = p*psi
    p_plus = np.fmax(p_plus,par.p_min) # lower bound
    p_plus = np.fmin(p_plus,par.p_max) # upper bound
    return p_plus

@njit(fastmath=True)
def n_plus_func(d,par):
    n_plus = (1-par.delta)*d
    n_plus = np.fmin(n_plus,par.n_max) # upper bound
    return n_plus

@njit(fastmath=True)
def n1_plus_func(d1,par):
    n1_plus = (1-par.delta1)*d1
    n1_plus = np.fmin(n1_plus,par.n_max) # upper bound
    return n1_plus

@njit(fastmath=True)
def n2_plus_func(d2,par):
    n2_plus = (1-par.delta2)*d2
    n2_plus = np.fmin(n2_plus,par.n_max) # upper bound
    return n2_plus

@njit(fastmath=True)
def m_plus_func(a,p_plus,xi_plus,par):
    y_plus = p_plus*xi_plus
    m_plus = par.R*a+ y_plus
    return m_plus

@njit(fastmath=True)
def x_plus_func(m_plus,n_plus,par):
    return m_plus + (1-par.tau)*n_plus